### ISSUE TITLE
Buttons drawn by wxRendererMac::DrawTreeItemButton don't respect dark mode

### ISSUE THREAD

**Issue by craftyjon** [wxWidgets/wxWidgets#24002]
### Description

On wxMac, the DrawTreeItemButton renderer always draws in light mode (i.e. dark strokes) even when the system is in dark mode.  I'm still trying to track down why this is, and if there's a simple fix, since the macOS APIs used are old and not really documented.

I've played around with adding a call to `HIThemeSetStroke` before `HIThemeDrawButton`, but this doesn't seem to actually change how the button is drawn (or I'm just doing it wrong).  For example (I've tried various different ThemeBrush colors with no effect)

```
ThemeBrush color = (flags & wxCONTROL_DISABLED) ? kThemeBrushButtonFrameInactive : kThemeBrushButtonFrameActive;
HIThemeSetStroke( color, nullptr, cgContext, kHIThemeOrientationNormal );
```

Perhaps the `HIThemeDrawButton` API just is hard-coded to draw in black, in which case if there is not a newer macOS API (I haven't found one yet) maybe wxOSX should just switch to a generic renderer for these buttons inside a tree control (and other places they are needed)

### Platform and version information

- wxWidgets version  you use: 3.2.3
- wxWidgets port  you use: wxOSX
- OS  and its version: macOS 13.5.1

**Comment by craftyjon** [comment]
I guess the "right" path is to move to a Cocoa NSButton in disclosure mode, but I don't know enough about macOS development / Objective-C to actually attempt that myself (especially since it seems like it might even call for a whole new file in osx/cocoa, and I don't know where to start

**Comment by vadz** [comment]
@csomor If you could please give a hint about how to fix this, it would be very welcome as I have no idea about this neither.

In the worst case, we'd have to use the generic function in the dark mode, as it should be less wrong than drawing the invisible native button...

**Comment by csomor** [comment]
> @csomor If you could please give a hint about how to fix this, it would be very welcome as I have no idea about this neither.

I'll look into it

**Comment by csomor** [comment]
ok, yes, the HIThemeDraw calls seem to respect at certain places the appearance mode, but for the disclosure button this does not work. To really clean this up we'd have to write a new renderer that eg uses NSCells that are allocated on a invisible NSWindow and convinces them to draw into the current cgContext. So for the moment I'd suggest using a generic implementation.

**Comment by craftyjon** [comment]
I tried doing:

```cpp
    // HIThemeDrawButton does not respect dark mode
    if ( wxSystemSettings::GetAppearance().IsDark() )
        return wxRendererNative::GetGeneric().DrawTreeItemButton( win, dc, rect, flags );
```

This works, but looks quite bad ("unnatural") on Mac, so in my opinion the Mac at least needs a new generic implementation that has an art style that matches the platform style

Generic look:
<img width="55" alt="image" src="https://github.com/wxWidgets/wxWidgets/assets/116737/13b590f7-61f4-4c4a-8774-ded279c2e31b">

Desired look:
<img width="86" alt="image" src="https://github.com/wxWidgets/wxWidgets/assets/116737/90ec5e91-0f39-4706-ab39-761f046d032d">

**Comment by csomor** [comment]
@craftyjon thanks for the samples, in the meantime I have done a proof of concept for a NSCellRenderer using NSCells for checkbox and disclosure buttons, I'll do a PR with that so that we can focus on this, I'll leave the old HITheme calls in there for all the others, so that we always have something that should work more or less...
@vadz apart from the render sample, do we have other apps to test against ? 
Thanks

**Comment by csomor** [comment]
see #24004 , you will have to manually adapt the makefiles, Xcode projects at this moment
instead of osx/carbon/renderer.cpp osx/cocoa/renderer.mm should be used

**Comment by vadz** [comment]
> @vadz apart from the render sample, do we have other apps to test against ?

Quite a few controls use `wxRendererNative`: grid, list, tree, propgrid etc.

> you will have to manually adapt the makefiles, Xcode projects at this moment
instead of osx/carbon/renderer.cpp osx/cocoa/renderer.mm should be used

Is the latter supposed to completely replace the former or do we still need to keep it for whatever reason?

**Comment by csomor** [comment]
> > @vadz apart from the render sample, do we have other apps to test against ?
> 
> Quite a few controls use `wxRendererNative`: grid, list, tree, propgrid etc.

I've already tested with the first three, I'll add propgrid, thanks alot

> > you will have to manually adapt the makefiles, Xcode projects at this moment
> > instead of osx/carbon/renderer.cpp osx/cocoa/renderer.mm should be used
> 
> Is the latter supposed to completely replace the former or do we still need to keep it for whatever reason?

no, it will be replaced completely by the .mm version

**Comment by vadz** [comment]
> I've already tested with the first three, I'll add propgrid, thanks a lot

For completeness, it's also used by `wxComboCtrl` (which can be seen in the `combo` sample) and `wxBitmapButton::CreateCloseButton()` (which is used by `wxInfoBar` and can be seen in the `dialogs` sample).

> > > you will have to manually adapt the makefiles, Xcode projects at this moment
> > > instead of osx/carbon/renderer.cpp osx/cocoa/renderer.mm should be used
> > 
> > Is the latter supposed to completely replace the former or do we still need to keep it for whatever reason?
> 
> no, it will be replaced completely by the .mm version

Please let me know if you want me to push the update to the makefiles to your PR in case you don't want to deal with bakefile yourself.

**Comment by csomor** [comment]
> Please let me know if you want me to push the update to the makefiles to your PR in case you don't want to deal with bakefile yourself.

thanks a lot for the offer, I think I made it work, after some fights getting all the ingredients ... 
btw I only found a release download for 0.2.12, 0.2.13 only as source, is this correct ? ... as building 0.2.13 from scratch wasn't possible and autoconf choked on missing files I added the ones missing from the 0.2.12, at least it's running bakefiles_gen again ...

**Comment by craftyjon** [comment]
This is fixed by 21da0e128 right?

**Comment by vadz** [comment]
Yes, I've just retested and it works fine for me, thanks for the reminder.
